File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/MobileLinkPolicy.kt
package com.lextures.android.core.lms

import java.net.URI

/**
 * Pure MB.1 link classification — mirrored on iOS. No networking / UI.
 * Uses [java.net.URI] so unit tests run on the JVM without Android stubs.
 */
object MobileLinkPolicy {
    enum class Handling(val wire: String) {
        IN_APP("in_app"),
        SYSTEM("system"),
        BLOCKED("blocked");

        companion object {
            fun parse(raw: String?): Handling = when (raw?.trim()?.lowercase()) {
                "system" -> SYSTEM
                "blocked" -> BLOCKED
                else -> IN_APP
            }
        }
    }

    enum class Classification(val wire: String) {
        NATIVE("native"),
        IN_APP_BROWSER("in_app_browser"),
        SYSTEM_BROWSER("system_browser"),
        EXTERNAL_APP("external_app"),
        AUTH_SESSION("auth_session"),
        BLOCKED("blocked"),
    }

    data class State(
        val handling: Handling = Handling.IN_APP,
        /** When false, external http(s) fall through to the system browser.
         *  Production always leaves this true; tests may force false. */
        val inAppBrowserEnabled: Boolean = true,
        /** Host of the configured API base (may include port). */
        val apiHost: String = "",
        val firstPartyHosts: List<String> = listOf("lextures.com", "localhost", "127.0.0.1"),
    )

    private val externalSchemes = setOf("mailto", "tel", "sms", "geo", "maps", "itms-apps", "market")
    private val webSchemes = setOf("http", "https")

    private val nativeAppHostSuffixes = listOf(
        "zoom.us", "zoom.com", "meet.google.com", "teams.microsoft.com", "teams.live.com",
    )

    private val checkoutHostSuffixes = listOf(
        "checkout.stripe.com", "billing.stripe.com", "pay.stripe.com",
    )

    private val ssoPathMarkers = listOf(
        "/oauth", "/oidc", "/saml", "/sso", "/auth/callback", "/login/oauth",
    )

    /** Pure classification of a URL string (absolute, relative `/…`, or `lextures://…`). */
    fun classify(urlString: String, state: State): Classification {
        val trimmed = urlString.trim()
        if (trimmed.isEmpty()) return Classification.BLOCKED

        if (trimmed.startsWith("/")) return Classification.NATIVE

        val lower = trimmed.lowercase()
        // Scheme without full parse for known custom schemes (URI may reject some).
        val scheme = schemeOf(trimmed)?.lowercase()
        if (scheme == null && !lower.startsWith("http://") && !lower.startsWith("https://")) {
            // bare relative already handled; unknown
            return Classification.BLOCKED
        }

        if (scheme == "lextures") return Classification.NATIVE
        if (scheme != null && scheme in externalSchemes) return Classification.EXTERNAL_APP
        if (scheme != null && scheme !in webSchemes) return Classification.BLOCKED

        val uri = parseUri(trimmed) ?: return Classification.BLOCKED
        val host = (uri.host ?: "").lowercase()
        if (host.isEmpty()) return Classification.BLOCKED

        if (isFirstParty(host, state)) return Classification.NATIVE
        if (isNativeAppHost(host)) return Classification.EXTERNAL_APP
        if (isCheckoutHost(host)) return Classification.SYSTEM_BROWSER

        val path = (uri.path ?: "").lowercase()
        if (isSSOPath(path)) return Classification.AUTH_SESSION

        return when (state.handling) {
            Handling.BLOCKED -> Classification.BLOCKED
            Handling.SYSTEM -> Classification.SYSTEM_BROWSER
            Handling.IN_APP ->
                if (state.inAppBrowserEnabled) Classification.IN_APP_BROWSER
                else Classification.SYSTEM_BROWSER
        }
    }

    /**
     * True when an Authorization bearer header may be attached for this request host.
     * Parsed-host equality / subdomain — never string-prefix on full URLs (FR-20).
     */
    fun shouldAttachBearer(requestHost: String?, apiHost: String): Boolean {
        val req = stripPort(requestHost.orEmpty()).lowercase()
        val api = stripPort(apiHost).lowercase()
        if (req.isEmpty() || api.isEmpty()) return false
        if (req == api) return true
        if (req.endsWith(".$api")) return true
        if (isLoopback(req) && isLoopback(api)) return true
        return false
    }

    fun displayHost(url: String): String {
        val host = (parseUri(url)?.host ?: "").lowercase()
        return if (host.startsWith("www.")) host.removePrefix("www.") else host
    }

    fun isSecure(url: String): Boolean =
        schemeOf(url)?.lowercase() == "https"

    /** Content-free telemetry payload (FR-25 / AC-20). */
    fun sanitizeTelemetry(raw: Map<String, String>): Map<String, String> {
        val allowed = setOf("source", "classification", "outcome", "errorClass")
        return raw.filterKeys { it in allowed }
    }

    private fun schemeOf(raw: String): String? {
        val idx = raw.indexOf(':')
        if (idx <= 0) return null
        val scheme = raw.substring(0, idx)
        if (!scheme.all { it.isLetter() || it == '+' || it == '.' || it == '-' }) return null
        return scheme
    }

    private fun parseUri(raw: String): URI? {
        return try {
            // java.net.URI is case-sensitive on scheme for some ops; normalize scheme casing.
            val scheme = schemeOf(raw)
            val normalized = if (scheme != null && scheme != scheme.lowercase()) {
                scheme.lowercase() + raw.substring(scheme.length)
            } else {
                raw
            }
            URI(normalized)
        } catch (_: Exception) {
            null
        }
    }

    private fun isFirstParty(host: String, state: State): Boolean {
        val h = host.lowercase()
        for (candidate in state.firstPartyHosts) {
            val c = candidate.lowercase()
            if (h == c || h.endsWith(".$c")) return true
        }
        val api = stripPort(state.apiHost).lowercase()
        if (api.isNotEmpty() && (h == api || h.endsWith(".$api"))) return true
        return false
    }

    private fun isNativeAppHost(host: String): Boolean {
        val h = host.lowercase()
        return nativeAppHostSuffixes.any { suffix -> h == suffix || h.endsWith(".$suffix") }
    }

    private fun isCheckoutHost(host: String): Boolean {
        val h = host.lowercase()
        return checkoutHostSuffixes.any { suffix -> h == suffix || h.endsWith(".$suffix") }
    }

    private fun isSSOPath(path: String): Boolean =
        ssoPathMarkers.any { path.contains(it) }

    private fun stripPort(host: String): String {
        val idx = host.indexOf(':')
        return if (idx >= 0) host.substring(0, idx) else host
    }

    private fun isLoopback(host: String): Boolean {
        val h = stripPort(host).lowercase()
        return h == "localhost" || h == "127.0.0.1" || h == "::1"
    }
}

File: clients/ios/Lextures/Core/LMS/MobileLinkPolicy.swift
import Foundation

/// Pure MB.1 link classification — mirrored on Android. No networking / UI.
enum MobileLinkPolicy {
    enum Handling: String, Equatable {
        case inApp = "in_app"
        case system
        case blocked

        static func parse(_ raw: String?) -> Handling {
            switch (raw ?? "").trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
            case "system": return .system
            case "blocked": return .blocked
            default: return .inApp
            }
        }
    }

    enum Classification: String, Equatable {
        case native
        case inAppBrowser = "in_app_browser"
        case systemBrowser = "system_browser"
        case externalApp = "external_app"
        case authSession = "auth_session"
        case blocked
    }

    struct State: Equatable {
        var handling: Handling = .inApp
        /// When false, external http(s) fall through to the system browser.
        /// Production always leaves this true; tests may force false.
        var inAppBrowserEnabled: Bool = true
        /// Host of the configured API base (may include port).
        var apiHost: String = ""
        var firstPartyHosts: [String] = ["lextures.com", "localhost", "127.0.0.1"]
    }

    private static let externalSchemes: Set<String> = [
        "mailto", "tel", "sms", "geo", "maps", "itms-apps", "market",
    ]

    private static let webSchemes: Set<String> = ["http", "https"]

    private static let nativeAppHostSuffixes: [String] = [
        "zoom.us", "zoom.com", "meet.google.com", "teams.microsoft.com", "teams.live.com",
    ]

    private static let checkoutHostSuffixes: [String] = [
        "checkout.stripe.com", "billing.stripe.com", "pay.stripe.com",
    ]

    private static let ssoPathMarkers: [String] = [
        "/oauth", "/oidc", "/saml", "/sso", "/auth/callback", "/login/oauth",
    ]

    // MARK: - Classify

    /// Pure classification of a URL string (absolute, relative `/…`, or `lextures://…`).
    static func classify(urlString: String, state: State) -> Classification {
        let trimmed = urlString.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return .blocked }

        // Relative in-app path.
        if trimmed.hasPrefix("/") {
            return .native
        }

        guard let components = URLComponents(string: trimmed) else {
            return .blocked
        }
        let scheme = (components.scheme ?? "").lowercased()

        if scheme == "lextures" {
            return .native
        }

        if externalSchemes.contains(scheme) {
            return .externalApp
        }

        if !webSchemes.contains(scheme) {
            // Unknown / dangerous schemes (javascript:, file:, data:, …).
            return .blocked
        }

        let host = (components.host ?? "").lowercased()
        if host.isEmpty { return .blocked }

        if isFirstParty(host: host, state: state) {
            return .native
        }

        if isNativeAppHost(host) {
            return .externalApp
        }

        if isCheckoutHost(host) {
            return .systemBrowser
        }

        let path = components.path.lowercased()
        if isSSOPath(path) {
            return .authSession
        }

        // External http(s) — honour admin policy + feature flag.
        switch state.handling {
        case .blocked:
            return .blocked
        case .system:
            return .systemBrowser
        case .inApp:
            return state.inAppBrowserEnabled ? .inAppBrowser : .systemBrowser
        }
    }

    // MARK: - Bearer attachment (FR-20 / FR-21)

    /// True when an Authorization bearer header may be attached for this request host.
    /// Uses parsed-host equality / subdomain — never string-prefix on full URLs.
    static func shouldAttachBearer(requestHost: String?, apiHost: String) -> Bool {
        let req = stripPort(requestHost ?? "").lowercased()
        let api = stripPort(apiHost).lowercased()
        if req.isEmpty || api.isEmpty { return false }
        if req == api { return true }
        // Subdomain of api host (e.g. cdn.api.lextures.com under api.lextures.com).
        if req.hasSuffix("." + api) { return true }
        // Dev: 127.0.0.1 / localhost treated as first-party API when api is loopback.
        if isLoopback(req) && isLoopback(api) { return true }
        return false
    }

    static func shouldAttachBearer(requestURL: URL, apiBaseURL: URL) -> Bool {
        shouldAttachBearer(requestHost: requestURL.host, apiHost: apiBaseURL.host ?? "")
    }

    // MARK: - Display helpers

    /// Registrable-ish host for chrome (drops leading www.).
    static func displayHost(for url: URL) -> String {
        let host = (url.host ?? "").lowercased()
        if host.hasPrefix("www.") { return String(host.dropFirst(4)) }
        return host
    }

    static func isSecure(_ url: URL) -> Bool {
        (url.scheme ?? "").lowercased() == "https"
    }

    // MARK: - Telemetry shape (FR-25 / AC-20)

    /// Returns a content-free telemetry payload, dropping any forbidden keys.
    static func sanitizeTelemetry(_ raw: [String: String]) -> [String: String] {
        let allowed: Set<String> = ["source", "classification", "outcome", "errorClass"]
        var out: [String: String] = [:]
        for (k, v) in raw where allowed.contains(k) {
            out[k] = v
        }
        return out
    }

    // MARK: - Internals

    private static func isFirstParty(host: String, state: State) -> Bool {
        let h = host.lowercased()
        for candidate in state.firstPartyHosts {
            let c = candidate.lowercased()
            if h == c || h.hasSuffix("." + c) { return true }
        }
        // API host itself is first-party for deep links that point at the API origin.
        let api = stripPort(state.apiHost).lowercased()
        if !api.isEmpty && (h == api || h.hasSuffix("." + api)) { return true }
        return false
    }

    private static func isNativeAppHost(_ host: String) -> Bool {
        let h = host.lowercased()
        for suffix in nativeAppHostSuffixes {
            if h == suffix || h.hasSuffix("." + suffix) { return true }
        }
        return false
    }

    private static func isCheckoutHost(_ host: String) -> Bool {
        let h = host.lowercased()
        for suffix in checkoutHostSuffixes {
            if h == suffix || h.hasSuffix("." + suffix) { return true }
        }
        return false
    }

    private static func isSSOPath(_ path: String) -> Bool {
        for marker in ssoPathMarkers {
            if path.contains(marker) { return true }
        }
        return false
    }

    private static func stripPort(_ host: String) -> String {
        if let idx = host.firstIndex(of: ":") {
            return String(host[..<idx])
        }
        return host
    }

    private static func isLoopback(_ host: String) -> Bool {
        let h = stripPort(host).lowercased()
        return h == "localhost" || h == "127.0.0.1" || h == "::1"
    }
}
